cdci_data_analysis/pytest_fixtures.py

```diff
@@ -644,6 +656,24 @@ def dispatcher_live_fixture(pytestconfig, dispatcher_test_conf_fn, dispatcher_de
                 
         kill_child_processes(pid, signal.SIGINT)
         os.kill(pid, signal.SIGINT)
+
+
+@pytest.fixture
+def second_dispatcher_live_fixture(pytestconfig, second_dispatcher_test_conf_fn, dispatcher_debug, request):
```

Comment:
A multi-threaded dispatcher probably would be the way, but it might take more development.

Also aren't we supposed to have load-balancer, that would assign the request to another dispatcher running instance? If so, in this way is closer to the situation we would have in production/staging

Reply:
> A multi-threaded dispatcher probably would be the way, but it might take more development.

But did you check if it's not already being multi-threaded? 

Also I do not think it's more development to make sure it's multi-threaded if it is not. It's just more involved development, you'd need to remember how the [dispatcher is started](https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/cdci_data_analysis/pytest_fixtures.py#L466).
 
> Also aren't we supposed to have load-balancer, that would assign the request to another dispatcher running instance? If so, in this way is closer to the situation we would have in production/staging

Dispatchers are running multi-threaded in production. There may be also several of them. But the complex situation occurs when the requests are to another thread, if it occurs, there may be some interaction between the requests. This is the situation which should be tested primarily.

Reply:
> 
> But did you check if it's not already being multi-threaded?
> 
> Also I do not think it's more development to make sure it's multi-threaded if it is not. It's just more involved development, you'd need to remember how the [dispatcher is started](https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/cdci_data_analysis/pytest_fixtures.py#L466).
> 

It is started as single threaded https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/bin/run_osa_cdci_server.py#L109

Has the dispatcher ever been used multi-threaded? Do we have experience with that?

> 
> Dispatchers are running multi-threaded in production. There may be also several of them. 

So does that mean that each dispatcher is started with the option `threaded=True` ?

Reply:
> > But did you check if it's not already being multi-threaded?
> > Also I do not think it's more development to make sure it's multi-threaded if it is not. It's just more involved development, you'd need to remember how the [dispatcher is started](https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/cdci_data_analysis/pytest_fixtures.py#L466).
> 
> It is started as single threaded
> 
> https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/bin/run_osa_cdci_server.py#L109
> 
> Has the dispatcher ever been used multi-threaded? Do we have experience with that?

 
> > Dispatchers are running multi-threaded in production. There may be also several of them.
> 
> So does that mean that each dispatcher is started with the option `threaded=True` ?

Not really, because it's started with gunicorn (of course, since flask server is no good for production). 
A variety of options were tried to optimize load.

I checked and actually, right now, we only run multiprocessing with gunicorn

https://github.com/oda-hub/dispatcher-container/blob/7c5ca47dcb643c85539e245f3109b534669a7bbd/entrypoint.sh#L24-L31

It might be easy to just use this command in the fixture.
Or test with threading, which is more demanding and will be useful.

Reply:
> 
> I checked and actually, right now, we only run multiprocessing with gunicorn
> 
> https://github.com/oda-hub/dispatcher-container/blob/7c5ca47dcb643c85539e245f3109b534669a7bbd/entrypoint.sh#L24-L31

Isn't this equivalent to the unicorn-based start within the dispatcher?

https://github.com/oda-hub/dispatcher-app/blob/c3d66dda0adf9cdc1ac589b63050f6ae5fee45d5/bin/run_osa_cdci_server.py#L89-L107

Reply:
I am testing with the options `threaded=False`, but it does not seem to be working fully here on github, no issue at all locally on my machine

Reply:
It is using now gunicorn